README.md
# connect jsonp

## Installation

    $ npm install connect-jsonp

## Examples

See included <a href='http://github.com/steelThread/connect-jsonp/blob/master/examples/app.js'>example</a>.

## Testing

    git submodule update --init
    make test

## TODOs

See <a href="http://github.com/steelThread/connect-jsonp/issues">issues</a> that begin with 'TODO:'

## License 

(The MIT License)

Copyright (c) 2010 Sean McDaniel

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md
0.0.2 / 2010-10-06
==================

  * Eking out efficiencies

0.0.1 / 2010-09-28
==================

  * Initial Release

test/connect-jsonp.test.js
var connect = require('connect'),
    jsonp = require('connect-jsonp'),
    helpers = require('./helpers'),
    assert = require('assert'),
	mime = require('connect/utils').mime;

connect.jsonp = jsonp;

var expectRes = "{'data': 'data'}";
var expectPaddedRes = "cb({'data': 'data'})";


function server(headers, code) {
    headers = headers || {};
    code = code || 200;
    return helpers.run(
        connect.gzip(),
        connect.jsonp(),
        connect.createServer(
            function(req, res) {
                res.writeHead(code, headers);
                res.write(expectRes)
                res.end();
            }
        )
    );
}

function assertPadded(res, padded) {
    if (padded) {
        assert.eql(expectPaddedRes, res.body);
        assert.eql(mime.type('.js'), res.headers['content-type']);
        assert.eql(expectPaddedRes.length, res.headers['content-length']);
    
    } else {
        assert.eql(expectRes, res.body);     
    }
}

module.exports = {
    'test not a padding request': function() {
        var req = server().request('GET', '/', { });
        req.buffer = true;
        req.addListener('response', function(res) {
            res.addListener('end', function() {
                assertPadded(res, false);
            });
        });
        req.end();
    },

    'test not a GET request': function() {
        var expect = 'cb({"error":"method not allowed","description":"with callback only GET allowed"})';
        helpers.run(connect.jsonp()).assertResponse(
            'POST', 
            '/?callback=cb', 
            400, 
            expect,
            '',
            function(res) {
                assert.eql(mime.type('.js'), res.headers['content-type']);
                assert.eql(expect.length, res.headers['content-length']);
            }
        );
    },

    'test query string': function() {
        var req = server().request('GET', '/test?callback=cb', { });
        req.buffer = true;
        req.addListener('response', function(res) {
            res.addListener('end', function() {
                assertPadded(res, true);
            });
        });
        req.end();  
    },


   'test plays nice with others': function() {
        var req = server().request('GET', '/test?callback=cb', {
            'Accept-Encoding': 'deflate, gzip'
        });
        req.buffer = true;
        req.addListener('response', function(res) {
            res.addListener('end', function() {
                assert.eql(mime.type('.js'), res.headers['content-type']);
                assert.eql('gzip', res.headers['content-encoding']);
                // test compression
                //assert.eql("cb({data: 'data'})", res.body);
            });
        });
        req.write('callback=cb');
        req.end();
    }
}

examples/app.js
// note: assumes the connect & connect-json node packages are installed 
var connect = require('connect'),
    jsonp = require('connect-jsonp');

var response = {
    success: true,
    it: 'works!'    
};

var server = connect.createServer(
    connect.logger({ format: ':method :url' }),
    connect.bodyDecoder(),
    jsonp(),
    connect.router(app),
    connect.errorHandler({ dumpExceptions: true, showStack: true })

).listen(3000);

console.log('connect-jsonp example server listening on port 3000');

function app(app) {
    // simple alert, not exactly x-domain...
    app.get('/', function(req, res) {
        res.writeHead(200, {'Content-Type': 'text/html'});
        res.end('<script type="text/javascript" src="/script-tag?callback=alert"></script>');
    });

    // requested after the script tag is rendered, result is evaluated
    app.get('/script-tag', function(req, res) {
        res.writeHead(200, {'Content-Type': 'application/json'});
        res.end(JSON.stringify(response));      
    });git
}


lib/connect-jsonp.js
/*!
 * Copyright(c) 2010 Sean McDaniel <sean.mcdaniel@me.com>
 * MIT Licensed
 */

/**
 * Simple connect middleware providing JSONP support.
 *
 * @return {Function}
 * @api public
 */
module.exports = function jsonp() { 
    // @@
    var APP_JS_CONTENT_TYPE = require('connect/utils').mime.type('.js');
    var BAD_REQUEST_BODY = '(' + 
        JSON.stringify({
            error: 'method not allowed', 
            description: 'with callback only GET allowed'       
        }) + 
    ')';

   /**
    * Write out a 400 error, passing it to the callback.
    */
   function badRequest(res, callback) {
       var body = callback + BAD_REQUEST_BODY;      
       res.writeHead(400, {
           'Content-Type': APP_JS_CONTENT_TYPE,
           'Content-Length': body.length
       });
       res.end(body);      
   };    

   /**
    * Unwind.
    */
   function previous(statusCode, headers, res, context, body, encoding) {
       res.writeHead = context.writeHead;
       res.writeHead(statusCode, headers); 
       if (body) {
           res.write = context.write;
           res.end = context.end;
           res.end(body, encoding);
       }
   };

    /**
     * Response decorator that pads a (json) response with a callback
     * if requested do to so.  
     */ 
    return function jsonp(req, res, next) {
        var url = require('url').parse(req.url, true);
        if (!(url.query && url.query.callback)) {
            next();
            return;
    
        } else if (req.method != 'GET') {
            badRequest(res, url.query.callback); 
            return;
        } 

        var context = {
            writeHead: res.writeHead,
            write: res.write,
            end: res.end
        };   

        res.writeHead = function(statusCode, headers) {
            if (statusCode === 200) {
                var body = '';
                res.write = function(chunk, encoding) { body += chunk }    
                res.end = function(chunk, encoding) {
                    if (chunk) body += chunk;
                    body = url.query.callback + '(' + body + ')';
                    headers['Content-Type'] = APP_JS_CONTENT_TYPE;
                    headers['Content-Length'] = body.length;
                    previous(statusCode, headers, res, context, body, encoding);
                };

            } else {
                previous(statusCode, headers, res, context);
            }   
        };

        next();
    };
};